src/librustc_codegen_ssa/mir/mod.rs

```diff
@@ -423,7 +427,27 @@ fn arg_local_refs<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
             bx.store_fn_arg(arg, &mut llarg_idx, tmp);
             LocalRef::Place(tmp)
         }
-    }).collect()
+    }).collect::<Vec<_>>();
+
+    if fx.instance.def.requires_caller_location(bx.tcx()) {
+        assert_eq!(
+            fx.fn_abi.args.len(), args.len() + 1,
+            "#[track_caller] fn's must have 1 more argument in their ABI than in their MIR",
+        );
+
+        let arg = &fx.fn_abi.args.last().unwrap();
+        match arg.mode {
+            PassMode::Direct(_) => (),
+            _ => panic!("caller location must be PassMode::Direct, found {:?}", arg.mode),
+        }
+
+        fx.caller_location = Some(OperandRef {
+            val: OperandValue::Immediate(bx.get_param(llarg_idx)),
+            layout: arg.layout,
```

Comment:
Not a problem, but I find it funny that codegen just doesn't use `caller_location_ty`, just gets it from elsewhere.

Reply:
Heh, good point. I think there's a class of bug that doing it this way catches, where the additional argument gets added at the wrong index but the length assertion and the pass mode are correct. If we synthesize the layout I think we'd have much stranger debugging to do than an ICE from a mismatched layout and use of it, right?